tf_encrypted/protocol/aby3/test_aby3.py

```diff
@@ -0,0 +1,1043 @@
+
+    a = tf.random_uniform([6, 10000], -3, 3)
+    b = tf.random_uniform([10000, 1], -0.1, 0.1)
+    x = tfe.define_private_input("input-provider", lambda: a)
+    y = tfe.define_private_input("input-provider", lambda: b)
+
+    z = tfe.matmul(x, y)
+    w = tf.matmul(a, b)
+
+    with tfe.Session() as sess:
+        sess.run(tfe.global_variables_initializer())
+        for i in range(1000):
+            result1, result2 = sess.run(
+                [z.reveal(), w])
+            close(result1, result2, 0.1)
+        print("test_error succeeds")
+
+
+def test_write_private():
+    tf.reset_default_graph()
+
+    prot = ABY3()
+    tfe.set_protocol(prot)
+
+    def provide_input():
+        # normal TensorFlow operations can be run locally
+        # as part of defining a private input, in this
+        # case on the machine of the input provider
+        return tf.ones(shape=(2, 2)) * 1.3
+
+    # define inputs
+    x = tfe.define_private_input('input-provider', provide_input)
+
+    write_op = x.write("x.tfrecord")
+
+    with tfe.Session() as sess:
+        # initialize variables
+        sess.run(tfe.global_variables_initializer())
+        # reveal result
+        sess.run(write_op)
+        print("test_write_private succeeds")
+
+
+def test_read_private():
+
+    tf.reset_default_graph()
+
+    prot = ABY3()
+    tfe.set_protocol(prot)
+
+    def provide_input():
+        return tf.reshape(tf.range(0, 8), [4, 2])
+
+    # define inputs
+    x = tfe.define_private_input('input-provider', provide_input)
+
+    write_op = x.write("x.tfrecord")
+
+    with tfe.Session() as sess:
+        # initialize variables
+        sess.run(tfe.global_variables_initializer())
+        # reveal result
+        sess.run(write_op)
+
+    x = tfe.read("x.tfrecord", batch_size=5, n_columns=2)
+    with tfe.Session() as sess:
+        result = sess.run(x.reveal())
+        close(result, np.array(list(range(0, 8)) + [0, 1]).reshape([5, 2]))
+        print("test_read_private succeeds")
+
+
+def test_iterate_private():
+    tf.reset_default_graph()
+
+    prot = ABY3()
+    tfe.set_protocol(prot)
+
+    def provide_input():
+        return tf.reshape(tf.range(0, 8), [4, 2])
+
+    # define inputs
+    x = tfe.define_private_input('input-provider', provide_input)
+
+    write_op = x.write("x.tfrecord")
+
+    with tfe.Session() as sess:
+        # initialize variables
+        sess.run(tfe.global_variables_initializer())
+        # reveal result
+        sess.run(write_op)
+
+    x = tfe.read("x.tfrecord", batch_size=5, n_columns=2)
+    y = tfe.iterate(x, batch_size=3, repeat=True, shuffle=False)
+    z = tfe.iterate(x, batch_size=3, repeat=True, shuffle=True)
+    with tfe.Session() as sess:
+        sess.run(tfe.global_variables_initializer())
+        print(sess.run(x.reveal()))
+        print(sess.run(y.reveal()))
+        print(sess.run(y.reveal()))
+        print(sess.run(x.reveal()))
+        print(sess.run(z.reveal()))
+
+
+if __name__ == "__main__":
```

Comment:
pytest seems to pick up these tests on its own so perhaps this listing of tests could be avoided altogether (or eg replaced by `unittest.main()` as in `pond_test.py`)? see https://github.com/tf-encrypted/tf-encrypted/issues/736